File: src/main/java/com/weeth/global/auth/resolver/CurrentUserRoleArgumentResolver.java
package com.weeth.global.auth.resolver;

import com.weeth.domain.user.domain.entity.enums.Role;
import com.weeth.global.auth.annotation.CurrentUserRole;
import com.weeth.global.auth.jwt.exception.AnonymousAuthenticationException;
import com.weeth.global.auth.model.AuthenticatedUser;
import org.springframework.core.MethodParameter;
import org.springframework.security.authentication.AnonymousAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.support.WebDataBinderFactory;
import org.springframework.web.context.request.NativeWebRequest;
import org.springframework.web.method.support.HandlerMethodArgumentResolver;
import org.springframework.web.method.support.ModelAndViewContainer;

public class CurrentUserRoleArgumentResolver implements HandlerMethodArgumentResolver {

    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        boolean hasAnnotation = parameter.hasParameterAnnotation(CurrentUserRole.class);
        boolean parameterType = Role.class.isAssignableFrom(parameter.getParameterType());
        return hasAnnotation && parameterType;
    }

    @Override
    public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer, NativeWebRequest webRequest, WebDataBinderFactory binderFactory) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();

        if (authentication == null || authentication instanceof AnonymousAuthenticationToken) {
            throw new AnonymousAuthenticationException();
        }

        Object principal = authentication.getPrincipal();
        if (principal instanceof AuthenticatedUser authenticatedUser) {
            return authenticatedUser.role();
        }

        for (GrantedAuthority authority : authentication.getAuthorities()) {
            String role = authority.getAuthority();
            if (role != null && role.startsWith("ROLE_")) {
                return Role.valueOf(role.substring("ROLE_".length()));
            }
        }

        throw new AnonymousAuthenticationException();
    }
}

File: src/main/kotlin/com/weeth/global/auth/resolver/CurrentUserRoleArgumentResolver.kt
package com.weeth.global.auth.resolver

import com.weeth.domain.user.domain.entity.enums.Role
import com.weeth.global.auth.annotation.CurrentUserRole
import com.weeth.global.auth.jwt.application.exception.AnonymousAuthenticationException
import com.weeth.global.auth.model.AuthenticatedUser
import org.springframework.core.MethodParameter
import org.springframework.security.authentication.AnonymousAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.bind.support.WebDataBinderFactory
import org.springframework.web.context.request.NativeWebRequest
import org.springframework.web.method.support.HandlerMethodArgumentResolver
import org.springframework.web.method.support.ModelAndViewContainer

class CurrentUserRoleArgumentResolver : HandlerMethodArgumentResolver {
    override fun supportsParameter(parameter: MethodParameter): Boolean {
        val hasAnnotation = parameter.hasParameterAnnotation(CurrentUserRole::class.java)
        val parameterType = Role::class.java.isAssignableFrom(parameter.parameterType)
        return hasAnnotation && parameterType
    }

    override fun resolveArgument(
        parameter: MethodParameter,
        mavContainer: ModelAndViewContainer?,
        webRequest: NativeWebRequest,
        binderFactory: WebDataBinderFactory?,
    ): Any {
        val authentication = SecurityContextHolder.getContext().authentication

        if (authentication == null || authentication is AnonymousAuthenticationToken) {
            throw AnonymousAuthenticationException()
        }

        val principal = authentication.principal
        if (principal is AuthenticatedUser) {
            return principal.role
        }

        val role =
            authentication.authorities
                .asSequence()
                .mapNotNull { authority -> authority.authority }
                .filter { it.startsWith("ROLE_") }
                .mapNotNull { raw ->
                    runCatching { Role.valueOf(raw.removePrefix("ROLE_")) }.getOrNull()
                }.firstOrNull()

        return role ?: throw AnonymousAuthenticationException()
    }
}
